### ISSUE TITLE
Design gadget API or interaction model

### ISSUE THREAD

**Issue by str4d** [zcash/halo2#64]
We will want some way to create reusable gadgets that can be composed to create halo2 circuits. This is pretty closely coupled to the general circuit API.

If possible, it would be great to maintain the bellman approach of keeping circuit generation and witness synthesis close together. The current halo2 circuit API doesn't quite achieve that (they are both part of the same trait but occur in different functions), so we either need to adjust that API or design gadgets to abstract over it.

**Comment by str4d** [comment]
We decided on a model that represents circuits as being composed of one or more "chips", which implement "instruction sets" that are required by "gadgets". A circuit then configures itself with chips that provide the set of necessary instructions, and then uses them in the gadgets it needs via a "layouter" that knows how to assign regions of cells within a circuit.

The main traits for this were implemented in #125. There may be some additional helper structs that we define as we go, but all that remains for this issue is to ensure the design is documented in the book.
